File: backend/src/main/java/com/example/focusquest/blocking/RulePrecedence.java
package com.example.focusquest.blocking;

import java.util.Collection;
import java.util.Comparator;
import java.util.Optional;

/**
 * Decides whether a URL is blocked given the active block and allowlist rules (mirrored by the
 * extension's {@code rulePrecedence.ts}).
 *
 * <p>Among all rules that match the URL, the most specific one wins (see
 * {@link UrlRule#compareTo}). When an allowlist rule and a block rule are equally specific, the
 * allowlist rule wins. This yields the documented order: most-specific allowlist, most-specific
 * block, broader allowlist, broader block, default allow.
 *
 * <p>Examples with block {@code example.com} and allow {@code example.com/docs}:
 * {@code example.com/docs/setup} is allowed and {@code example.com/forum} is blocked.
 */
public final class RulePrecedence {

    public enum Verdict {
        BLOCKED,
        ALLOWED_BY_ALLOWLIST,
        ALLOWED_BY_DEFAULT
    }

    /**
     * @param verdict     the outcome
     * @param matchedRule the rule that decided it; empty for {@link Verdict#ALLOWED_BY_DEFAULT}
     */
    public record Decision(Verdict verdict, Optional<UrlRule> matchedRule) {

        public boolean isBlocked() {
            return verdict == Verdict.BLOCKED;
        }
    }

    private RulePrecedence() {
    }

    public static Decision evaluate(TargetUrl url, Collection<UrlRule> blockRules, Collection<UrlRule> allowRules) {
        Optional<UrlRule> allow = mostSpecificMatch(url, allowRules);
        Optional<UrlRule> block = mostSpecificMatch(url, blockRules);

        if (allow.isPresent() && (block.isEmpty() || allow.get().compareTo(block.get()) >= 0)) {
            return new Decision(Verdict.ALLOWED_BY_ALLOWLIST, allow);
        }
        if (block.isPresent()) {
            return new Decision(Verdict.BLOCKED, block);
        }
        return new Decision(Verdict.ALLOWED_BY_DEFAULT, Optional.empty());
    }

    private static Optional<UrlRule> mostSpecificMatch(TargetUrl url, Collection<UrlRule> rules) {
        return rules.stream()
                .filter(rule -> rule.matches(url))
                .max(Comparator.naturalOrder());
    }
}

File: extension/src/blocking/rulePrecedence.ts
// Decides whether a URL is blocked given the active block and allowlist rules. Reproduces the
// backend's RulePrecedence and UrlRule.compareTo.
//
// Among all rules that match the URL, the most specific one wins. When an allowlist rule and a
// block rule are equally specific, the allowlist rule wins. That yields the documented order:
// most-specific allowlist, most-specific block, broader allowlist, broader block, default allow.
//
// Example with block `example.com` and allow `example.com/docs`: `example.com/docs/setup` is
// allowed and `example.com/forum` is blocked.

import type { Decision, TargetUrl, UrlRule } from '../types/blocking'
import { ruleMatches } from './urlMatcher'

export function hostLabelCount(rule: UrlRule): number {
  return rule.host.split('.').length
}

export function pathSegmentCount(rule: UrlRule): number {
  return rule.path.split('/').length - 1
}

/**
 * Orders rules by specificity: more host labels first (`m.youtube.com` beats `youtube.com`), then
 * more path segments (`/shorts/a` beats `/shorts`). Positive when `a` is more specific than `b`.
 */
export function compareSpecificity(a: UrlRule, b: UrlRule): number {
  return hostLabelCount(a) - hostLabelCount(b) || pathSegmentCount(a) - pathSegmentCount(b)
}

const MAX_SCORED_SEGMENTS = 99

/**
 * A single number that orders rules the same way as {@link compareSpecificity}, for the
 * declarativeNetRequest rule `priority`. Host labels outrank path segments; paths deeper than
 * 99 segments (far beyond the 253-character rule limit's practical use) are treated as equal.
 */
export function specificityScore(rule: UrlRule): number {
  return hostLabelCount(rule) * (MAX_SCORED_SEGMENTS + 1) + Math.min(pathSegmentCount(rule), MAX_SCORED_SEGMENTS)
}

function mostSpecificMatch(url: TargetUrl, rules: readonly UrlRule[]): UrlRule | null {
  let best: UrlRule | null = null
  for (const rule of rules) {
    if (ruleMatches(rule, url) && (best === null || compareSpecificity(rule, best) > 0)) best = rule
  }
  return best
}

export function evaluate(url: TargetUrl, blockRules: readonly UrlRule[], allowRules: readonly UrlRule[]): Decision {
  const allow = mostSpecificMatch(url, allowRules)
  const block = mostSpecificMatch(url, blockRules)

  if (allow && (!block || compareSpecificity(allow, block) >= 0)) {
    return { verdict: 'ALLOWED_BY_ALLOWLIST', matchedRule: allow, isBlocked: false }
  }
  if (block) return { verdict: 'BLOCKED', matchedRule: block, isBlocked: true }
  return { verdict: 'ALLOWED_BY_DEFAULT', matchedRule: null, isBlocked: false }
}
